In [1]:
import numpy as np

In [2]:
import pandas as pd

In [3]:
movies = pd.read_csv('tmdb_5000_movies.csv')

In [4]:
credits = pd.read_csv('tmdb_5000_credits.csv')

In [5]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [6]:
credits.head(1)

movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [8]:
movies.shape

(4803, 20)

In [9]:
movies = movies.merge(credits, on = 'title')

In [11]:
movies.shape

(4809, 23)

In [12]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

In [15]:
movies = movies[['movie_id' , 'title', 'overview' , 'genres' , 'keywords' , 'cast', 'crew']]

In [16]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4809 entries, 0 to 4808
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4809 non-null   int64 
 1   title     4809 non-null   object
 2   overview  4806 non-null   object
 3   genres    4809 non-null   object
 4   keywords  4809 non-null   object
 5   cast      4809 non-null   object
 6   crew      4809 non-null   object
dtypes: int64(1), object(6)
memory usage: 300.6+ KB


In [17]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [18]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [19]:
movies.dropna(inplace = True)

C:\Users\hp\AppData\Local\Temp\ipykernel_3228\3024812739.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies.dropna(inplace = True)


In [21]:
movies.isnull().sum()

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [22]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

# '[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'
convert to 

['Action' , 'Adventure', 'Fantasy' , 'SciFi']

In [27]:
import ast

def convert(obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

In [28]:
movies['genres'] = movies['genres'].apply(convert)

What we did here? Using ast ,we converted the object to list , and then among those took only the name of the genres and appended to the list , and in the end returned List , and finally stored List L in the movies['genres'] = L

In [29]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [30]:
movies['keywords'] = movies['keywords'].apply(convert)

In [31]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [32]:
movies['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [37]:
import ast

def convert3(obj):
    L = []
    counter = 0;
    for i in ast.literal_eval(obj):
        if(counter!=3):
            L.append(i['name'])
            counter+=1
        else:
            break
        
    return L

In [39]:
movies['cast'] = movies['cast'].apply(convert3)

In [40]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [42]:
def fetch_director(obj):
    L= []
    for i in ast.literal_eval(obj):
        if(i['job'] == 'Director'):
            L.append(i['name'])
            break
    return L

In [43]:
movies['crew'] = movies['crew'].apply(fetch_director)

In [45]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

In [46]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [48]:
movies['overview'] = movies['overview'].apply(lambda x: x.split())
##We converted the sen in to words

In [49]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

# removing space , since it might lead to confusion in recommendation 

In [52]:
movies['genres'].apply(lambda x : [i.replace(" " , "") for i in x])
movies['keywords'].apply(lambda x : [i.replace(" " , "") for i in x])
movies['cast'].apply(lambda x : [i.replace(" " , "") for i in x])
movies['crew'].apply(lambda x : [i.replace(" " , "") for i in x])

0           [JamesCameron]
1          [GoreVerbinski]
2              [SamMendes]
3       [ChristopherNolan]
4          [AndrewStanton]
               ...        
4804     [RobertRodriguez]
4805         [EdwardBurns]
4806          [ScottSmith]
4807          [DanielHsia]
4808     [BrianHerzlinger]
Name: crew, Length: 4806, dtype: object

In [54]:
movies['tag'] = movies['overview']+movies['genres']+movies['keywords']+movies['cast']+movies['crew']

In [55]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]   

                                                 tag  
0  [In, the, 22nd, century,, a, paraplegic, Marin...

In [57]:
new_df = movies[['movie_id' , 'title' , 'tag']]

In [58]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                 tag  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [61]:
new_df['tag'] = new_df['tag'].apply(lambda x : " ".join(x))

C:\Users\hp\AppData\Local\Temp\ipykernel_3228\1149718486.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tag'] = new_df['tag'].apply(lambda x : " ".join(x))


In [62]:
new_df['tag'] = new_df['tag'].apply(lambda x : x.lower())

C:\Users\hp\AppData\Local\Temp\ipykernel_3228\1711057993.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tag'] = new_df['tag'].apply(lambda x : x.lower())


# NOW WE WILL DO TEXT VECTORIZATION

# We will use BOW 

In [69]:
from sklearn.feature_extraction.text import CountVectorizer


In [70]:
cv = CountVectorizer(max_features = 5000 , stop_words = 'english')

In [75]:
vectors = cv.fit_transform(new_df['tag']).toarray()

In [76]:
vectors[1]

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

# Apply stemming to avoid same words

In [80]:
import nltk

In [81]:
from nltk.stem.porter import PorterStemmer

In [82]:
ps = PorterStemmer()

In [86]:
def stem(text):
    y = []
    for i in text.split():
        y.append(ps.stem(i))
        
    return " ".join(y)
        

In [88]:
new_df['tag'] = new_df['tag'].apply(stem)

C:\Users\hp\AppData\Local\Temp\ipykernel_3228\4074900442.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tag'] = new_df['tag'].apply(stem)


# Cosine distance is found out in this project 
angle between two vectors is cosine distance 
small cosine distance , means similar movies

In [89]:
from sklearn.metrics.pairwise import cosine_similarity

In [93]:
similarity = cosine_similarity(vectors)

In [94]:
similarity

array([[1.        , 0.06885304, 0.04948717, ..., 0.03142697, 0.05410018,
        0.        ],
       [0.06885304, 1.        , 0.04259177, ..., 0.04057204, 0.        ,
        0.        ],
       [0.04948717, 0.04259177, 1.        , ..., 0.01944039, 0.08924215,
        0.        ],
       ...,
       [0.03142697, 0.04057204, 0.01944039, ..., 1.        , 0.06375767,
        0.03276488],
       [0.05410018, 0.        , 0.08924215, ..., 0.06375767, 1.        ,
        0.03760222],
       [0.        , 0.        , 0.        , ..., 0.03276488, 0.03760222,
        1.        ]])

In [96]:
similarity[0]
#This is the similarity of first movie with all the other movies 

array([1.        , 0.06885304, 0.04948717, ..., 0.03142697, 0.05410018,
       0.        ])

In [135]:
def recommend(movie):
    movie_index = new_df[new_df['title'] == movie].index[0]
    distances = similarity[movie_index]
    ## we need to store the index and movie so we will use ennumerate function 
    movies_list = sorted(list(enumerate(distances)) , reverse = True , key = lambda x : x[1])[1:6]
    for i in movies_list:
        print (new_df.iloc[i[0]].title)
        

        

# Explanation of the above function 
Here movie_index = new_df[new_df['title'] == movie].index[0]
will fetch the index of the movie 

distances = similarity[movie_index]
corresponding to the given index it will return the distance

movies_list = sorted(list(enumerate(distances)) , reverse = True , key = lambda x : x[1])[1:6]
This function will sort the distance according to the second value and we have used enumerate function to prevent loss of ordering of movies 

print(i[0]) ->this will print the movies id 

print(new_df.iloc[i[0]].title)
=> this will print the movie name corresponding to the given id 

In [131]:
similarity[0]

array([1.        , 0.06885304, 0.04948717, ..., 0.03142697, 0.05410018,
       0.        ])

In [132]:
sorted(list(enumerate(similarity[0])) , reverse = True , key = lambda x : x[1])[1:6]

##reverse = True =>means we need sorting in reverse order

## key => means we need to sort the data using x[1] , means the second value

[(2409, 0.4839354795704658),
 (1537, 0.4335521648830699),
 (3162, 0.4278406467922595),
 (838, 0.398216086792002),
 (4336, 0.39207842352784267)]

In [136]:
recommend('Avatar')

Aliens
Moonraker
Alien
Alien³
Silent Running


In [128]:
new_df.iloc[2409].title


'Aliens'